internal/expr/expr_test.go

```diff
@@ -293,281 +233,195 @@ var tests = []struct {
-		structs:  []any{Address{ID: 1}},
-		errorstr: `cannot prepare expression: type Address has no "road" db tag`,
+		query:       "SELECT (p.*, t.name) AS &Address.* FROM t",
+		prepareArgs: []any{Address{}},
+		err:         "cannot prepare expression: invalid asterisk in output expression: (p.*, t.name) AS &Address.*",
 	}, {
-		sql:      "SELECT &Address.road FROM t",
-		structs:  []any{Address{ID: 1}},
-		errorstr: `cannot prepare expression: type Address has no "road" db tag`,
-	}}
-
-	for i, test := range testList {
-		parser := expr.NewParser()
-		parsedExpr, err := parser.Parse(test.sql)
-		if err != nil {
-			c.Fatal(err)
-		}
-		_, err = parsedExpr.Prepare(test.structs...)
-		c.Assert(err, ErrorMatches, test.errorstr,
-			Commentf("test %d failed:\nsql: '%s'\nstructs:'%+v'", i, test.sql, test.structs))
-
-	}
-}
-
-func (s *ExprSuite) TestPrepareInvalidAsteriskPlacement(c *C) {
-	testList := []struct {
-		sql      string
-		structs  []any
-		errorstr string
-	}{{
-		sql:      "SELECT (&Person.*, &Person.*) FROM t",
-		structs:  []any{Address{}, Person{}},
-		errorstr: "cannot prepare expression: invalid asterisk in output expression: Output[[] [Person.* Person.*]]",
+		query:       "SELECT (name, p.*) AS (&Person.id, &Person.*) FROM t",
+		prepareArgs: []any{Address{}, Person{}},
+		err:         "cannot prepare expression: invalid asterisk in output expression: (name, p.*) AS (&Person.id, &Person.*)",
 	}, {
-		sql:      "SELECT (p.*, t.*) AS &Address.* FROM t",
-		structs:  []any{Address{}},
-		errorstr: "cannot prepare expression: invalid asterisk in output expression: Output[[p.* t.*] [Address.*]]",
+		query:       "SELECT (&Person.*, &Person.*) FROM t",
+		prepareArgs: []any{Address{}, Person{}},
+		err:         "cannot prepare expression: invalid asterisk in output expression: (&Person.*, &Person.*)",
 	}, {
-		sql:      "SELECT p.* AS &Address.street FROM t",
-		structs:  []any{Address{}},
-		errorstr: "cannot prepare expression: invalid asterisk in output expression: Output[[p.*] [Address.street]]",
-	}}
-
-	for i, test := range testList {
-		parser := expr.NewParser()
-		parsedExpr, err := parser.Parse(test.sql)
-		if err != nil {
-			c.Fatal(err)
-		}
-		_, err = parsedExpr.Prepare(test.structs...)
-		c.Assert(err.Error(), Equals, test.errorstr,
-			Commentf("test %d failed:\nsql: '%s'\nstructs:'%+v'", i, test.sql, test.structs))
-	}
-}
-
-func (s *ExprSuite) TestPrepareAsteriskMix(c *C) {
-	testList := []struct {
-		sql      string
-		structs  []any
-		errorstr string
-	}{{
-		sql:      "SELECT (&Address.*, &Address.id) FROM t",
-		structs:  []any{Address{}, Person{}},
-		errorstr: "cannot prepare expression: invalid asterisk in output expression: Output[[] [Address.* Address.id]]",
+		query:       "SELECT (p.*, t.*) AS &Address.* FROM t",
+		prepareArgs: []any{Address{}},
+		err:         "cannot prepare expression: invalid asterisk in output expression: (p.*, t.*) AS &Address.*",
```

Comment:
This improves it significantly, but this error is still not great. No need to fix it now, but the error message gives no clue about what's actually wrong. Imagine this with an expression that is 5 lines long.

Reply:
Hopefully nobody will be writing a 5 line long output expression, but yea, point taken.